qiskit/pulse/ir/pulse_ir.py

```diff
@@ -0,0 +1,400 @@
+
+        Recursively flattens all child ``PulseIR`` until only instructions remain."""
+        new_elements = self._get_flatten_elements()
+        self._elements = new_elements
+
+    def _get_flatten_elements(self) -> List[Union[BaseIRInstruction, "PulseIR"]]:
+        """Return a list of flatten instructions
+
+        Recursively lists all instructions in the ``PulseIR``.
+
+        Raises:
+            PulseError: If any instruction in the ``PulseIR`` is not scheduled
+                (has ``initial_time==None``)
+        """
+        flatten_elements = []
+        for element in self._elements:
+            if isinstance(element, PulseIR):
+                flatten_elements.extend(element._get_flatten_elements())
+            else:
+                if element.initial_time is None:
+                    raise PulseError("Can not flatten an unscheduled PulseIR")
+                flatten_elements.append(element)
+        return flatten_elements
+
+    def get_instructions_by_mixed_frame(
+        self,
+        mixed_frame: MixedFrame,
+        recursive: Optional[bool] = True,
+    ) -> List[GenericInstruction]:
+        """Return all instructions associated with a given ``MixedFrame``
+
+        Args:
+            mixed_frame: The ``MixedFrame`` whose instructions are to be returned.
+            recursive (Optional): If ``True`` recursively looks for instructions,
+                else ignores instructions of child ``PulseIR``. Default - ``True``.
+        """
+
+        instructions = []
+        for element in self._elements:
+            if isinstance(element, GenericInstruction):
+                if mixed_frame == MixedFrame(element.logical_element, element.frame):
+                    instructions.append(element)
+            elif recursive and isinstance(element, PulseIR):
+                instructions.extend(
+                    element.get_instructions_by_mixed_frame(mixed_frame, recursive=recursive)
+                )
+        return instructions
+
+    def get_acquire_instructions(
+        self, qubit: Optional[Qubit] = None, recursive: Optional[bool] = True
+    ) -> List[GenericInstruction]:
+        """Return ``AcquireInstruction``\\ s
+
+        Args:
+            qubit: Optionally return only instructions associated with this qubit.
+                Default value - None (return all instructions).
+            recursive: If ``True`` recursively looks for instructions,
+                else ignores instructions of child ``PulseIR``. Default - ``True``.
+        """
+        instructions = []
+        for element in self._elements:
+            if isinstance(element, AcquireInstruction) and (
+                qubit is None or element.qubit == qubit
+            ):
+                instructions.append(element)
+            elif recursive and isinstance(element, PulseIR):
+                instructions.extend(element.get_acquire_instructions(qubit, recursive=recursive))
+        return instructions
```

Comment:
It seems like we should provide generic means for getting instructions, not fixed routines for each function?